smac_planner/src/smac_planner.cpp

```diff
@@ -454,6 +483,65 @@ void SmacPlanner::removeHook(std::vector<Eigen::Vector2d> & path)
   }
 }
 
+void SmacPlanner::initCostmap()
+{
+  if (_downsample_costmap && _resolution > _original_costmap->getResolution()) {
+    _costmap = _costmap_downsampler->downsample(_original_costmap, _resolution);
+    createOccupancyGridFromCostmap(_costmap);
+    _downsampled_costmap_pub->publish(std::move(_grid));
+  } else {
+    _costmap = _original_costmap;
+  }
+}
+
+  void SmacPlanner::createOccupancyGridFromCostmap(const nav2_costmap_2d::Costmap2D * costmap)
+{
+  float grid_resolution = costmap->getResolution();
+  float grid_width = costmap->getSizeInCellsX();
+  float grid_height = costmap->getSizeInCellsY();
+
+  _grid = std::make_unique<nav_msgs::msg::OccupancyGrid>();
+
+  _grid->header.frame_id = _global_frame;
+  _grid->header.stamp = rclcpp::Time();
+
+  _grid->info.resolution = grid_resolution;
+
+  _grid->info.width = grid_width;
+  _grid->info.height = grid_height;
+
+  double wx, wy;
+  costmap->mapToWorld(0, 0, wx, wy);
+  _grid->info.origin.position.x = wx - grid_resolution / 2;
+  _grid->info.origin.position.y = wy - grid_resolution / 2;
+  _grid->info.origin.position.z = 0.0;
+  _grid->info.origin.orientation.w = 1.0;
+
+  _grid->data.resize(_grid->info.width * _grid->info.height);
+
+  unsigned char * data = costmap->getCharMap();
+  for (unsigned int i = 0; i < _grid->data.size(); i++)
+  {
+    _grid->data[i] = _cost_translation_table[data[i]];
+  }
+}
+
+void SmacPlanner::setCostTranslationTable() {
```

Comment:
Does something like this not exist in the costmap_2d layered costmap we have access to?

Reply:
You're right, I can use directly a new instance of https://github.com/SteveMacenski/navigation2/blob/nav2_smac_planner/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_publisher.hpp instead of just borrowing from that class what I need.